pkg/action/executor/move_util.go

```diff
@@ -97,22 +121,76 @@ func movePod(client *kclient.Clientset, pod *api.Pod, nodeName string, retryNum
 		}
 	}()
 
-	//2 wait until podC gets ready
-	err = podutil.WaitForPodReady(client, npod.Namespace, npod.Name, nodeName,
+	// Special handling for pods with volumes
+	if podUsingVolume {
+		// We still support replicaset and replication controllers as parents,
+		// however both of them could further have a parent (deploy or deploy config).
+		parent, grandParent, pClient, gPClient, gPKind, err :=
+			getParentAndGrandParentInfo(clusterScraper, pod, parentKind)
+		if err != nil {
+			return nil, err
+		}
+
+		if grandParent != nil {
+			pause := true
+			unpause := false
+			// step B8: (via defer)
+			defer func() {
+				glog.V(3).Infof("Unpausing pods controller: %s for pod: %s", gPKind, podQualifiedName)
+				resourceRollout(gPClient, grandParent, unpause)
+			}()
+
+			// step B2:
+			glog.V(3).Infof("Pausing pods controller: %s for pod: %s", gPKind, podQualifiedName)
+			err := resourceRollout(gPClient, grandParent, pause)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		valid := true
+		invalid := false
+		// step B7: (via defer)
+		defer func() {
+			glog.V(3).Infof("Updating scheduler of %s for pod %s back to default.", parentKind, podQualifiedName)
+			changeScheduler(pClient, parent, valid)
+
+		}()
+		// step B3:
+		glog.V(3).Infof("Updating scheduler of %s for pod %s to unknown (turbo-scheduler).", parentKind, pod.Name)
+		changeScheduler(pClient, parent, invalid)
+
+		// step A3/B4:
+		// We optimistically delete the original pod (PodA), so that the new pod can
+		// bind to the volume.
+		// TODO: This is not an ideal way of doing things, and users should be
+		// encouraged to rather disable move actions on pods which use volumes
+		// via a config either in kubeturbo or driven from server UI.
+		delOpt := &metav1.DeleteOptions{}
+		if err := podClient.Delete(pod.Name, delOpt); err != nil {
+			glog.Errorf("Move pod warning: failed to delete original pod: %v", err)
+			return nil, err
+		}
+	}
+
+	//step A2/B5: wait until podC gets ready
+	err = podutil.WaitForPodReady(clusterScraper.Clientset, npod.Namespace, npod.Name, nodeName,
```

Comment:
I can think of a few scenarios that this might run into problems (although chances are rare):

* If the previous Delete of the original pod does not complete for whatever reason (e.g., stuck in Terminating), then the cloned pod won't start. After timeout, the cloned pod is deleted, and scheduler/rollout is reenabled, at which time, a new pod will be scheduled to bump up the replica, but the new pod will probably still fail to start because the original pod is stuck in Terminating
* The original pod is deleted successfully, but the cloned pod fails to start for whatever reason. After timeout, the cloned pod is deleted, and scheduler/rollout is reenabled, at which time, a new pod will be scheduled to bump up the replica, but the new pod could run anywhere in the cluster.

I do not have a good solution (if you agree that the above can happen), but at least we can keep them in mind in case we run into them and we can give an explanation.

Reply:
As discussed, I will compile a proactive readme which can list possible corner cases, which although rare can happen.